In [1]:
## Standard Imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
## Importing the OS and JSON Modules
import os,json

In [2]:
df = pd.read_csv('superhero_info - superhero_info.csv')
df.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}  
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}

In [3]:
df1 = pd.read_csv('superhero_powers - superhero_powers.csv')
df1.head()

hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

## 1. Separate 'Hero' and 'Publisher' into two columns.

In [4]:
# Exploring existing format 
df['Hero|Publisher']



0               A-Bomb|Marvel Comics
1       Abe Sapien|Dark Horse Comics
2                 Abin Sur|DC Comics
3          Abomination|Marvel Comics
4        Absorbing Man|Marvel Comics
                   ...              
458       Yellowjacket|Marvel Comics
459    Yellowjacket II|Marvel Comics
460                Yoda|George Lucas
461                Zatanna|DC Comics
462                   Zoom|DC Comics
Name: Hero|Publisher, Length: 463, dtype: object

In [5]:
## adding expand=True
df['Hero|Publisher'].str.split('|',expand=True)

0                  1
0             A-Bomb      Marvel Comics
1         Abe Sapien  Dark Horse Comics
2           Abin Sur          DC Comics
3        Abomination      Marvel Comics
4      Absorbing Man      Marvel Comics
..               ...                ...
458     Yellowjacket      Marvel Comics
459  Yellowjacket II      Marvel Comics
460             Yoda       George Lucas
461          Zatanna          DC Comics
462             Zoom          DC Comics

[463 rows x 2 columns]

In [6]:
## save the 2 new columns into the dataframe
df[['Hero','Publisher']] = df['Hero|Publisher'].str.split('|',expand=True)
df.head(2)

Hero|Publisher Gender           Race Alignment Hair color  \
0          A-Bomb|Marvel Comics   Male          Human      good    No Hair   
1  Abe Sapien|Dark Horse Comics   Male  Icthyo Sapien      good    No Hair   

  Eye color Skin color                                  Measurements  \
0    yellow    Unknown  {'Height': '203.0 cm', 'Weight': '441.0 kg'}   
1      blue       blue   {'Height': '191.0 cm', 'Weight': '65.0 kg'}   

         Hero          Publisher  
0      A-Bomb      Marvel Comics  
1  Abe Sapien  Dark Horse Comics

In [7]:
## drop the original column 
df = df.drop(columns=['Hero|Publisher'])
df.head(2)



Gender           Race Alignment Hair color Eye color Skin color  \
0   Male          Human      good    No Hair    yellow    Unknown   
1   Male  Icthyo Sapien      good    No Hair      blue       blue   

                                   Measurements        Hero          Publisher  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}      A-Bomb      Marvel Comics  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  Abe Sapien  Dark Horse Comics

## 2. Unpack the 'Measurements' column of dictionaries into separate columns

- We will to convert the single "Measurements" column into 2 separate columns "Height" and "Weight".

- First we will replace the single quotes in string-dictionaries with double quotes so we can use json.loads

In [8]:
import json
## use .str.replace to replace all single quotes
df['Measurements'] = df['Measurements'].str.replace("'",'"')
## Apply the json.loads to the full column
df['Measurements'] = df['Measurements'].apply(json.loads)
df['Measurements'].head()

0    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
1     {'Height': '191.0 cm', 'Weight': '65.0 kg'}
2     {'Height': '185.0 cm', 'Weight': '90.0 kg'}
3    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
4    {'Height': '193.0 cm', 'Weight': '122.0 kg'}
Name: Measurements, dtype: object

In [9]:
height_weight = df['Measurements'].apply(pd.Series)
height_weight

Height    Weight
0    203.0 cm  441.0 kg
1    191.0 cm   65.0 kg
2    185.0 cm   90.0 kg
3    203.0 cm  441.0 kg
4    193.0 cm  122.0 kg
..        ...       ...
458  183.0 cm   83.0 kg
459  165.0 cm   52.0 kg
460   66.0 cm   17.0 kg
461  170.0 cm   57.0 kg
462  185.0 cm   81.0 kg

[463 rows x 2 columns]

## 3. Change the titles names of 'Height' and 'Weight' columns to 'Height(cm)' and 'Weight(kg)' and then remove 'cm' and 'kg' from the values in each column so we can have only numbers.

In [10]:
## save the 2 new columns into the dataframe with new title names
df[['Height (cm)','Weight (kg)']] = df['Measurements'].apply(pd.Series)
df.head()


Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher Height (cm) Weight (kg)  
0      Marvel Comics    203.0 cm    441.0 kg  
1  Dark Horse Comics    191.0 cm     65.0 kg  
2          DC Comics    185.0 cm     90.0 kg  
3      Marvel Comics    203.0 cm    441.0 kg  
4      Marvel Comics    193.0 cm    122.0 kg

In [11]:
# Make a list of all characters to replace
to_replace = ['cm','kg']
# run a loop to replace all of the characters in the list at once
for char in to_replace:
    df['Height (cm)'] = df['Height (cm)'].str.replace(char,'',regex=False)
    df['Weight (kg)'] = df['Weight (kg)'].str.replace(char,'',regex=False)
df.head()


Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher Height (cm) Weight (kg)  
0      Marvel Comics      203.0       441.0   
1  Dark Horse Comics      191.0        65.0   
2          DC Comics      185.0        90.0   
3      Marvel Comics      203.0       441.0   
4      Marvel Comics      193.0       122.0

Let's drop the original 'Measurements' column

In [12]:
df = df.drop(columns=['Measurements'])
df.head()


Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

            Hero          Publisher Height (cm) Weight (kg)  
0         A-Bomb      Marvel Comics      203.0       441.0   
1     Abe Sapien  Dark Horse Comics      191.0        65.0   
2       Abin Sur          DC Comics      185.0        90.0   
3    Abomination      Marvel Comics      203.0       441.0   
4  Absorbing Man      Marvel Comics      193.0       122.0

In [13]:
# Check to see if the 'Height' and 'Weight' columns are numeric
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Gender       463 non-null    object
 1   Race         463 non-null    object
 2   Alignment    463 non-null    object
 3   Hair color   463 non-null    object
 4   Eye color    463 non-null    object
 5   Skin color   463 non-null    object
 6   Hero         463 non-null    object
 7   Publisher    463 non-null    object
 8   Height (cm)  463 non-null    object
 9   Weight (kg)  463 non-null    object
dtypes: object(10)
memory usage: 36.3+ KB


In [14]:
# Change the 'Height' and 'Weight' columns to float64 datatype
df['Height (cm)'] = df['Height (cm)'].astype('float64')
df['Weight (kg)'] = df['Weight (kg)'].astype('float64')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Gender       463 non-null    object 
 1   Race         463 non-null    object 
 2   Alignment    463 non-null    object 
 3   Hair color   463 non-null    object 
 4   Eye color    463 non-null    object 
 5   Skin color   463 non-null    object 
 6   Hero         463 non-null    object 
 7   Publisher    463 non-null    object 
 8   Height (cm)  463 non-null    float64
 9   Weight (kg)  463 non-null    float64
dtypes: float64(2), object(8)
memory usage: 36.3+ KB


- The 'Height' and 'Weight' columns are now floats which is a numeric datatype.

## 4. Convert the 'Powers' column to list

In [15]:
## showing how the powers appear in the 'Powers' column
df1.loc[2,'Powers']


'Agility,Accelerated Healing,Cold Resistance,Durability,Underwater breathing,Marksmanship,Weapons Master,Longevity,Intelligence,Super Strength,Telepathy,Stamina,Immortality,Reflexes,Enhanced Sight,Sub-Mariner'

In [16]:
# Create a new column that has the 'Powers' column converted to a list 
df1['Powers_split'] = df1['Powers'].str.split(",")

In [17]:
df1.loc[2,'Powers_split']

['Agility',
 'Accelerated Healing',
 'Cold Resistance',
 'Durability',
 'Underwater breathing',
 'Marksmanship',
 'Weapons Master',
 'Longevity',
 'Intelligence',
 'Super Strength',
 'Telepathy',
 'Stamina',
 'Immortality',
 'Reflexes',
 'Enhanced Sight',
 'Sub-Mariner']

In [18]:
df1.head()

hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

                                        Powers_split  
0    [Agility, Super Strength, Stamina, Super Speed]  
1  [Accelerated Healing, Durability, Longevity, S...  
2  [Agility, Accelerated Healing, Cold Resistance...  
3                               [Lantern Power Ring]  
4  [Accelerated Healing, Intelligence, Super Stre...

## 5. Convert the 'Powers' column into one-hot-encoded columns (a column for each item in the list)

In [19]:
df1['Powers_split'].value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 5231, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[Intelligence]                                                                                                                                                                                                                                                                          8
[Durability, Super Strength]                                                                                                                                                                                                                                                            5
[Agility, Stealth, Marksmanship, Weapons Master, Stamina]                                                                                                                                                                                                                               4
[Marksmanship]                                                                                                                                            

- We first need to get a list of all of the unique options that appear in this column.


- We will then loop through these unique options and leverage .str.contains to create a new column for each. 

In [20]:
## exploding the column of lists
exploded = df1.explode('Powers_split')
exploded[['hero_names','Powers', 'Powers_split']].head()

hero_names                                             Powers  \
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   

          Powers_split  
0              Agility  
0       Super Strength  
0              Stamina  
0          Super Speed  
1  Accelerated Healing

In [21]:
## saving the unique values from the exploded column
cols_to_make = exploded['Powers_split'].dropna().unique()
cols_to_make

array(['Agility', 'Super Strength', 'Stamina', 'Super Speed',
       'Accelerated Healing', 'Durability', 'Longevity', 'Camouflage',
       'Self-Sustenance', 'Cold Resistance', 'Underwater breathing',
       'Marksmanship', 'Weapons Master', 'Intelligence', 'Telepathy',
       'Immortality', 'Reflexes', 'Enhanced Sight', 'Sub-Mariner',
       'Lantern Power Ring', 'Invulnerability', 'Animation',
       'Super Breath', 'Dimensional Awareness', 'Flight', 'Size Changing',
       'Teleportation', 'Magic', 'Dimensional Travel',
       'Molecular Manipulation', 'Energy Manipulation', 'Power Cosmic',
       'Energy Absorption', 'Elemental Transmogrification',
       'Fire Resistance', 'Natural Armor', 'Heat Resistance',
       'Matter Absorption', 'Regeneration', 'Stealth', 'Power Suit',
       'Energy Blasts', 'Energy Beams', 'Heat Generation', 'Danger Sense',
       'Phasing', 'Force Fields', 'Hypnokinesis', 'Invisibility',
       'Enhanced Senses', 'Jump', 'Shapeshifting', 'Elasticity',
 

We will now use a For Loop and .str.contains to create the new columns

In [22]:
for col in cols_to_make:
    df1[col] = df1['Powers'].str.contains(col)
df1


C:\Users\tijes\anaconda3\envs\dojo-env\lib\site-packages\ipykernel_launcher.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  


hero_names                                             Powers  \
0            3-D Man         Agility,Super Strength,Stamina,Super Speed   
1             A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2         Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3           Abin Sur                                 Lantern Power Ring   
4        Abomination  Accelerated Healing,Intelligence,Super Strengt...   
..               ...                                                ...   
662  Yellowjacket II                 Flight,Energy Blasts,Size Changing   
663             Ymir  Cold Resistance,Durability,Longevity,Super Str...   
664             Yoda  Agility,Stealth,Danger Sense,Marksmanship,Weap...   
665          Zatanna  Cryokinesis,Telepathy,Magic,Fire Control,Proba...   
666             Zoom  Super Speed,Intangibility,Time Travel,Time Man...   

                                          Powers_split  Agility  \
0      [Agility, Super Strength, Stamina, Super Speed]     True   
1    [Accelerated Healing, Durability, Longevity, S...    False   
2    [Agility, Accelerated Healing, Cold Resistance...     True   
3                                 [Lantern Power Ring]    False   
4    [Accelerated Healing, Intelligence, Super Stre...    False   
..                                                 ...      ...   
662             [Flight, Energy Blasts, Size Changing]    False   
663  [Cold Resistance, Durability, Longevity, Super...    False   
664  [Agility, Stealth, Danger Sense, Marksmanship,...     True   
665  [Cryokinesis, Telepathy, Magic, Fire Control, ...    False   
666  [Super Speed, Intangibility, Time Travel, Time...    False   

     Super Strength  Stamina  Super Speed  Accelerated Healing  Durability  \
0              True     True         True                False       False   
1              True     True        False                 True        True   
2              True     True        False                 True        True   
3             False    False        False                False       False   
4              True     True         True                 True       False   
..              ...      ...          ...                  ...         ...   
662           False    False        False                False       False   
663            True    False        False                False        True   
664           False     True         True                False       False   
665           False    False        False                False       False   
666           False    False         True                False       False   

     Longevity  ...  Weather Control  Omnipresent  Omniscient  \
0        False  ...            False        False       False   
1         True  ...            False        False       False   
2         True  ...            False        False       False   
3        False  ...            False        False       False   
4        False  ...            False        False       False   
..         ...  ...              ...          ...         ...   
662      False  ...            False        False       False   
663       True  ...            False        False       False   
664       True  ...            False        False       False   
665      False  ...             True        False       False   
666      False  ...            False        False       False   

     Hair Manipulation  Nova Force  Odin Force  Phoenix Force  \
0                False       False       False          False   
1                False       False       False          False   
2                False       False       False          False   
3                False       False       False          False   
4                False       False       False          False   
..                 ...         ...         ...            ...   
662              False       False       False          False   
663              False       False       False          False   
6

- Now that we have one-hot-encoded columns for every power that appears in the dataset, we can delete the following columns: 'hero_names', 'Powers' and 'Powers_split'

In [23]:
df1 = df1.drop(columns=['Powers', 'Powers_split'])
df1.head()

hero_names  Agility  Super Strength  Stamina  Super Speed  \
0      3-D Man     True            True     True         True   
1       A-Bomb    False            True     True        False   
2   Abe Sapien     True            True     True        False   
3     Abin Sur    False           False    False        False   
4  Abomination    False            True     True         True   

   Accelerated Healing  Durability  Longevity  Camouflage  Self-Sustenance  \
0                False       False      False       False            False   
1                 True        True       True        True             True   
2                 True        True       True       False            False   
3                False       False      False       False            False   
4                 True       False      False       False            False   

   ...  Weather Control  Omnipresent  Omniscient  Hair Manipulation  \
0  ...            False        False       False              False   
1  ...            False        False       False              False   
2  ...            False        False       False              False   
3  ...            False        False       False              False   
4  ...            False        False       False              False   

   Nova Force  Odin Force  Phoenix Force  Intuitive aptitude  Melting  \
0       False       False          False               False    False   
1       False       False          False               False    False   
2       False       False          False               False    False   
3       False       False          False               False    False   
4       False       False          False               False    False   

   Changing Armor  
0           False  
1           False  
2           False  
3           False  
4           False  

[5 rows x 168 columns]

In [24]:
## merging with different column names
merged = pd.merge(df, df1, left_on='Hero', right_on='hero_names')
merged

Gender               Race Alignment        Hair color Eye color  \
0      Male              Human      good           No Hair    yellow   
1      Male      Icthyo Sapien      good           No Hair      blue   
2      Male            Ungaran      good           No Hair      blue   
3      Male  Human / Radiation       bad           No Hair     green   
4      Male              Human       bad           No Hair      blue   
..      ...                ...       ...               ...       ...   
458    Male              Human      good             Blond      blue   
459  Female              Human      good  Strawberry Blond      blue   
460    Male     Yoda's species      good             White     brown   
461  Female              Human      good             Black      blue   
462    Male            Unknown       bad             Brown       red   

    Skin color             Hero          Publisher  Height (cm)  Weight (kg)  \
0      Unknown           A-Bomb      Marvel Comics        203.0        441.0   
1         blue       Abe Sapien  Dark Horse Comics        191.0         65.0   
2          red         Abin Sur          DC Comics        185.0         90.0   
3      Unknown      Abomination      Marvel Comics        203.0        441.0   
4      Unknown    Absorbing Man      Marvel Comics        193.0        122.0   
..         ...              ...                ...          ...          ...   
458    Unknown     Yellowjacket      Marvel Comics        183.0         83.0   
459    Unknown  Yellowjacket II      Marvel Comics        165.0         52.0   
460      green             Yoda       George Lucas         66.0         17.0   
461    Unknown          Zatanna          DC Comics        170.0         57.0   
462    Unknown             Zoom          DC Comics        185.0         81.0   

     ... Weather Control  Omnipresent  Omniscient  Hair Manipulation  \
0    ...           False        False       False              False   
1    ...           False        False       False              False   
2    ...           False        False       False              False   
3    ...           False        False       False              False   
4    ...           False        False       False              False   
..   ...             ...          ...         ...                ...   
458  ...           False        False       False              False   
459  ...           False        False       False              False   
460  ...           False        False       False              False   
461  ...            True        False       False              False   
462  ...           False        False       False              False   

     Nova Force  Odin Force  Phoenix Force  Intuitive aptitude  Melting  \
0         False       False          False               False    False   
1         False       False          False               False    False   
2         False       False          False               False    False   
3         False       False          False               False    False   
4         False       False          False               False    False   
..          ...         ...            ...                 ...      ...   
458       False       False          False               False    False   
459       False       False          False               False    False   
460       False       False          False               False    False   
461       False       False          False               False    False   
462       False       False          False               False    False   

     Changing Armor  
0             False  
1             False  
2             False  
3             False  
4             False  
..              ...  
458           False  
459           False  
460           False  
461           False  
462           False  

[463 rows x 178 columns]

## 6. Use your combined DataFrame to answer the following questions.

a. Compare the average weight of super powers who have Super Speed to those who do not.

b. What is the average height of heroes for each publisher?

### a. Compare the average weight of super powers who have Super Speed to those who do not.

In [25]:
Power_weight = merged.groupby('Super Speed')['Weight (kg)'].mean()
Power_weight

Super Speed
False    101.773585
True     129.404040
Name: Weight (kg), dtype: float64

- The 'True' value above represents the average weight of super powers who have Super Speed. The 'False' value represents the average weight of super powers who do not have Super Speed.

- We can see that the average weight (129.40 kg) of super powers who have Super Speed is greater/higher than the average weight (101.77 kg) of super powers who do not have Super Speed.

### b. What is the average height of heroes for each publisher?

In [26]:
hero_height = merged.groupby('Publisher')['Height (cm)'].mean()
hero_height

Publisher
DC Comics            181.923913
Dark Horse Comics    176.909091
George Lucas         159.600000
Image Comics         211.000000
Marvel Comics        191.546128
Shueisha             171.500000
Star Trek            181.500000
Team Epic TV         180.750000
Unknown              178.000000
Name: Height (cm), dtype: float64

- The above has the average height (in cm) of heroes for each publisher. The Publisher is listed on the left side and the average heights on the right side.